controllers/core_clusteroperator_controller.go

```diff
@@ -0,0 +1,105 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	openshiftconfigv1 "github.com/openshift/api/config/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logr "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/openshift/platform-operators/internal/clusteroperator"
+)
+
+type CoreClusterOperatorReconciler struct {
+	client.Client
+	ReleaseVersion  string
+	SystemNamespace string
+}
+
+const (
+	coreCOName = "platform-operators-core"
+)
+
+//+kubebuilder:rbac:groups=platform.openshift.io,resources=platformoperators,verbs=list
+//+kubebuilder:rbac:groups=config.openshift.io,resources=clusteroperators,verbs=get;list;watch
+//+kubebuilder:rbac:groups=config.openshift.io,resources=clusteroperators/status,verbs=update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
+func (r *CoreClusterOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logr.FromContext(ctx)
+	log.Info("reconciling request", "req", req.NamespacedName)
+	defer log.Info("finished reconciling request", "req", req.NamespacedName)
+
+	coBuilder := clusteroperator.NewBuilder()
+	coWriter := clusteroperator.NewWriter(r.Client)
+
+	core := &openshiftconfigv1.ClusterOperator{}
+	if err := r.Get(ctx, req.NamespacedName, core); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+
+		log.Info("core clusteroperator does not exist. recreating it...", "name", coreCOName)
+		core = &openshiftconfigv1.ClusterOperator{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: coreCOName,
+			},
+			Status: openshiftconfigv1.ClusterOperatorStatus{},
+		}
+		return ctrl.Result{}, r.Create(ctx, core)
+	}
+	defer func() {
+		if err := coWriter.UpdateStatus(ctx, core, coBuilder.GetStatus()); err != nil {
+			log.Error(err, "error updating CO status")
+		}
+	}()
+
+	// Set the default ClusterOperator conditions to Available=True as the POM isn't
+	// managing any operands during phase 0.
+	coBuilder.WithDegraded(openshiftconfigv1.ConditionFalse)
+	coBuilder.WithAvailable(openshiftconfigv1.ConditionTrue, fmt.Sprintf("The platform operator manager is available at %s", r.ReleaseVersion), clusteroperator.ReasonAsExpected)
```

Comment:
Should one or both of Degraded/Available rely on any smoke tests? (e.g. can I list POs without error?)

Reply:
It depends on what smoke tests we think are reasonable. Reading through https://coreos.slack.com/archives/C030W4V9BHN/p1663354445071699 it seems like setting A=T is sufficient during phase 0 but I'm open to ideas.

Reply:
At a minimum, I think we should attempt a list of POs before we say A=T? I'm trying to think through various POM failures. Here's my stab at filling in a table.

| Failure | platform-operators-core | platform-operators-aggregated |
| ------------- | ------------- | --------------- |
| Can't communicate with apiserver | Available=F  | Available=F |
| Don't have permission to list POs  | Available=F  | Available=F |
| Failed to list bundles from catalog source(s) | Available=F | Available=F |
| Failed to list BDs | Available=F | Available=F |
| Failed to locate PO packageName in listed bundles | Available=T | Available=F |
| Don't have permission to create/update BDs for POs | Available=F | Available=F |
| BD for a PO did not progress to `InstallationSucceeded` | Available=T | Available=F |

>  Failed to list bundles from catalog source(s)

This one seems iffy since I recall that catsrcs GRPC connections can be flaky. Seems like we'd want to do retries for a little bit, then go `Degraded=true`, then eventually go `Available=false`
 
Also from the CVO docs:

> An operator shoould not report the Available status condition the first time until they are completely rolled out (or within some reasonable percentage if the component must be installed to all nodes)

This seems _super_ important to make sure we get right when managing PO installations for POs that are declared at cluster install time. I put #40 on hold while we think through this.

Reply:
And the value of listing the PlatformOperators in the cluster is to ensure that CRD has been registered? We basically only care about whether an error was returned vs. length checking the number of items the list query returned. I think my main concern is whether that opens the door for transient events, e.g. the API server is temporarily down, and whether toggling A=[T,F] is desirable in such a configuration.

I need to really dive more into the ClusterOperator documentation to get a better feel for what's applicable here given we don't explicitly manage any operands with this CO controller. Thoughts @bparees @wking?

Reply:
checking that you can list the api you care about is definitely more "aggressive" than i think most COs bother with.  That doesn't make it wrong, per se.

It accomplishes two things:

1) ensures the api is defined/available
2) ensures the operator itself has RBAC permissions to interact with the api

That said, i don't know that you need to explicitly poll it, more like if the control loop gets an error interacting with the api, then you might go degraded or unavailable.

Reply:
> This one seems iffy since I recall that catsrcs GRPC connections can be flaky. Seems like we'd want to do retries for a little bit, then go Degraded=true, then eventually go Available=false

How would we enforce this? My instinct is that this is overkill for phase 0, and we should try to avoid relying on the catalogsource gRPC connection unless we're implementing some sort of reference-counting-esq implementation. Does this also imply we'd need to combine the core ClusterOperator controller with the POM to be able to set CO status conditions during reconciliation?

> That said, i don't know that you need to explicitly poll it, more like if the control loop gets an error interacting with the api, then you might go degraded or unavailable.

Okay so in this implementation that checks whether we can successfully list the POs in the cluster, we'd be setting a degraded status vs. an unavailable status?

Reply:
> Okay so in this implementation that checks whether we can successfully list the POs in the cluster, we'd be setting a degraded status vs. an unavailable status?

if you can't list them, i'd say you're unavailable.  But you don't want to go unavailable just because you fail once.  If you fail once, you're degraded.  If you continue to fail for (5? 10? 15? 30?) minutes, you're unavailable.

Reply:
Do we examples of core components handling this kind of logic? It doesn't seem too difficult to implement how long we've been unable to perform a list query to determine unavailability, but it may involve defining some arbitrary toleration (e.g. 10m) for what we decide is unavailable, and may influence behaviors on other openshift distributions.

Reply:
i don't know of one offhand (we more commonly use the pattern on setting Alerts).

Honestly 5 mins should be plenty.  If we can't resolve the api resource after 5 minutes, nothing is likely to change that's going to fix it.

Not sure why it would have different implications on another openshift distribution, i can't think of any where we'd expect the behavior to be different.

Reply:
Briefly thinking about this some more: I wonder whether we need to distinguish between a "list failure threshold" before we mark the ClusterOperator as degraded. During reconciliation, if we fail to successfully perform a list query 3-5 in a row, then we mark the component as degraded. And then we'd also have a 5m threshold that distinguishes between D=T and A=T.

We're definitely introducing some complexity with this approach, so open to suggestions on whether this is a bad idea. I think my preference is still to keep the core ClusterOperator reconciliation loop pretty stupid during phase 0, and revisit this interaction as a GA requirement once we get a better picture of the overall OLM 1.x stack and our CVO integration model.

Reply:
absolutely not a phase 0 requirement.  

as for whether we need a threshold before going degraded...  i could see us just keeping it simple and not going degraded(at least not based on this).  Just have a threshold before you go unavailable.  As long as the threshold is 5 minutes or whatever, i think that's fine.  There's not much benefit to knowing a few minutes earlier "hey this thing is degraded"

So i may have talked us in circles, but ultimately i think "hey, i can't reach the apis i need to do my job, and i've been trying for 5 minutes, so i'm reporting myself unavailable" is probably sufficient.

Reply:
I can take a stab at this 5m threshold approach to determine A=[T,F] when I'm back on normal dev duties next week.

I'm mainly concerned with the initial cluster rollout edge case, and having a controller reconciliation loop that blindly sets A=T could be sufficient given we can assume the controller pod is running. Obviously it's not the most robust solution but helps side step any forward looking QE bug where we roll out a bad component container image during phase 0.

Reply:
> having a controller reconciliation loop that blindly sets A=T could be sufficient given we can assume the controller pod is running.

i think that is the right starting point.

Reply:
> as for whether we need a threshold before going degraded... i could see us just keeping it simple and not going degraded

I was thinking about this some more today: what happens if we're currently expressing A=T, and we fail a single list query? Ideally, we wouldn't set A=F given it could be transient issue. Would it make sense instead to set D=T here, which would update the LastTransistionTime, and keep the A=T and P=T conditions? Any subsequent reconciliations that continue to fail a list query would check whether D=T is already set, and if so, compare the LastTransistionTime to the failure threshold to determine whether we should be setting A=F.

@bparees @joelanford Thoughts? It was unclear to me reading through the states listed in https://github.com/openshift/enhancements/blob/master/dev-guide/cluster-version-operator/dev/clusteroperator.md#conditions-and-installupgrade whether this was a supported configuration.

Reply:
Took a rough stab at this implementation in https://github.com/openshift/platform-operators/pull/62/commits/cdda9b5e7a77dda2965d768942d67b835afc14e8.

Reply:
@timflannagan not sure what you mean by "supported configuration" but the algorithm/determination seems reasonable to me.  Of course you also need to clear D=T once you have a successful list, if and only if the `reason` for the D=T is failure to list.  (i.e. if you were degraded for some other reason, then successfully listing doesn't clear the degraded condition)